src/ui/components/SecondaryToolbox/NewConsole.tsx

```diff
@@ -0,0 +1,388 @@
+
+  // Update derived analysis points in a transition (so it's safe to Suspend) when points change.
+  useEffect(() => {
+    startTransition(() => {
+      setDeferredPoints(points);
+    });
+  }, [points]);
+
+  const context = useMemo(
+    () => ({
+      isPending,
+      points,
+      pointsForAnalysis: deferredPoints,
+
+      // PointsContext is read-only in this context.
+      // Log points are added by the legacy source Editor component.
+      addPoint: () => {},
+      deletePoint: () => {},
+      editPoint: () => {},
+    }),
+    [deferredPoints, isPending, points]
+  );
+
+  return <PointsContext.Provider value={context}>{children}</PointsContext.Provider>;
+}
+
+// Adapter that reads console evaluations (from Redux) and passes them to the TerminalContext.
+function TerminalContextReduxAdapter({ children }: PropsWithChildren) {
+  const [isPending, startTransition] = useTransition();
+  const [messages, setMessages] = useState<TerminalExpression[]>([]);
+
+  const dispatch = useAppDispatch();
+
+  const clearMessages = useCallback(() => {
+    dispatch(messagesClearEvaluations);
+    setMessages([]);
+  }, [dispatch]);
+
+  // Update derived terminal expressions in a transition (so it's safe to Suspend) when they change.
+  useEffect(() => {
+    const onEvaluation = (data: EvaluationEventPayload) => {
+      startTransition(() => {
+        setMessages(prevMessages => [
+          ...prevMessages,
+          {
+            ...data,
+            type: "TerminalExpression",
+          },
+        ]);
+      });
+    };
+
+    ThreadFront.on("evaluation", onEvaluation);
+    return () => {
+      ThreadFront.off("evaluation", onEvaluation);
+    };
+  }, []);
+
+  const context = useMemo(
+    () => ({
+      clearMessages,
+      isPending,
+      messages: messages,
+
+      // New terminal expressions can only be added via JSTerm (and Redux) in this context.
+      addMessage: () => {},
+    }),
+    [clearMessages, isPending, messages]
+  );
+
+  return <TerminalContext.Provider value={context}>{children}</TerminalContext.Provider>;
+}
+
+// Adapter that reads the current execution point and time (from Redux) and passes them to the TimelineContext.
+function TimelineContextAdapter({ children }: PropsWithChildren) {
+  const [state, setState] = useState<Omit<TimelineContextType, "isPending" | "update">>({
+    executionPoint: null,
+    pauseId: null,
+    time: 0,
+  });
+
+  const [isPending, startTransition] = useTransition();
+
+  const dispatch = useAppDispatch();
+  const pauseId = useAppSelector(getPauseId);
+  const time = useAppSelector(getCurrentTime);
+  const executionPoint = useAppSelector(getCurrentPoint);
+
+  const update = useCallback(
+    (time: number, executionPoint: ExecutionPoint, pauseId: PauseId) => {
+      dispatch(seek(executionPoint, time, false /* hasFrames */, pauseId));
+    },
+    [dispatch]
+  );
+
+  useLayoutEffect(() => {
```

Comment:
What's the reason for using `useLayoutEffect` here?

Reply:
It's probably not strictly _necessary_ but it avoids a browser paint in between where the new Console and the legacy app might reflect different points. (React flushes layout effects, and any scheduled updates, synchronously– before returning to the browser to paint).